### PR TITLE
Bug 1837677: open the VNC console in a new window

### PR DESCRIPTION
**Pull request by glekner**
This enables opening the vnc console in a new window.

**If the main window context is changed (e.g. move from console to details tab) then the new window will close aswell.**

Screens:
Screen Shot 2020-05-24 at 18 12 57

-----------------------

Screen Shot 2020-05-24 at 18 13 20

### CODE REVIEW

File: frontend/public/components/utils/window-portal.jsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import * as React from 'react';
```
Comment (glekner) [inline frontend/public/components/utils/window-portal.jsx:1]: @rawagner can you review :)

**Comment by yaacov** [comment]
@glekner Wow Great !

cc:// @jelkosz

**Comment by yaacov** [comment]
@glekner great :cake: I liked better the option of the new window with no buttons at all ( just the terminal ) can you make a screenshot of that option too ?

**Comment by yfrimanm** [comment]
This is great @glekner ! 

Please take a look at this design suggestion I have for rearranging the buttons
https://imgur.com/QDr2nMT

Thanks.
Yifat

**Comment by glekner** [comment]
- Applied @yfrimanm suggestion
- If Window creation fails, an alert is shown @yaacov

Screen Shot 2020-05-25 at 12 33 44

File: frontend/public/components/utils/window-portal.jsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  children,
+}) => {
+  const [containerEl] = React.useState(document.createElement('div'));
+  const externalWindow = React.useRef(null);
+
+  React.useEffect(
+    () => {
+      externalWindow.current = window.open(
+        '',
+        '',
```
Comment (yaacov) [inline frontend/public/components/utils/window-portal.jsx:37]: can you add the vmi name + "Console" as the window name ?

**Comment by suomiy** [comment]
Looks pretty good it has few limitations which could fixed. 

> If the main window context is changed (e.g. move from console to details tab) then the new window will close aswell.

IMO this is very counter-intuitive and we would get complaints about this soon.

cons
 - you need to have a host window and not use it in the future. Not until you close the terminal
 - you can't reference the terminal URL in case you need a quick look at the VM (more clicks have to be done)

We should have referenceable URL and proper standalone window for this to be user friendly.

Luckily something similar has already been done and is used in the console ([che workspaces terminal](https://github.com/openshift/console/pull/4762)):



they also have an option to open the terminal in standalone window at `my.domain/terminal`


this is how the init step looks (see the PR for other examples):

bb

can we get inspired by this and do something similar? And have our own url with the vm name like: `/k8s/ns/default/virtualmachines/my-example-vm/standalone-console` ?

**Comment by yaacov** [comment]
Notes:

> can we get inspired by this and do something similar? And have our own url with the vm name like: /k8s/ns/default/virtualmachines/my-example-vm/standalone-console ?

This is kubevirt backend endpoint for vnc [1]
`/apis/subresources.kubevirt.io/v1alpha3/namespaces/NAMESPACE/virtualmachineinstances/VM/vnc`
[1] https://kubevirt.io/2019/Access-Virtual-Machines-graphic-console-using-noVNC.html

**Comment by suomiy** [comment]
> This is kubevirt backend endpoint for vnc [1]
> `/apis/subresources.kubevirt.io/v1alpha3/namespaces/NAMESPACE/virtualmachineinstances/VM/vnc`

yes we are using that internally - although our console is also for serial and rdp. 

Do you want to encode the selected console into the url? Like `/default/virtualmachines/my-vm/consoles?console=vnc`

also, not sure we should be changing  vm -> vmi endpoint, because the consoles page should be also accessible when the vmi is non existent

**Comment by glekner** [comment]
> Do you want to encode the selected console into the url? Like `/default/virtualmachines/my-vm/consoles?console=vnc`

sounds like a good idea. do we need to handle permissions to those namespaces or they will already be dealt with?

**Comment by suomiy** [comment]
> sounds like a good idea. do we need to handle permissions to those namespaces or they will already be dealt with?

+1, we could have the same approach like the details page with StatusBox and show the appropriate error or `Restricted Access` message. Although we should offer redirect back to the console in that case.

**Comment by glekner** [comment]
Closing in favour of #5593

**Pull request closed without merging**